110 - Ported search to use typeorm

cowmanjoe opened the pull request:
The search API now uses typeorm as an ORM for querying and representing the data as typed objects. This also includes the changes I made for #99 and #94 (added filtering parameters).

This PR can be tested locally by putting your Postgres info inside the ormconfig.json. CAREFUL WHEN DOING THIS because if the synchronize option is on in the ormconfig.json, running it automatically edits the job table to match exactly the schema in job.ts, deleting the data from altered columns. This likely should be turned off in production to avoid deleting data, but it is convenient for testing locally.

cowmanjoe committed:
Added changes for min salary to search controller.

cowmanjoe committed:
Reverted changes to package-lock for client.

cowmanjoe committed:
Added filtering by radius to a point.

cowmanjoe committed:
Merge branch 'develop' into feature/99

cowmanjoe committed:
WIP using typeORM for server project.

cowmanjoe committed:
More work in progress.

cowmanjoe committed:
Fixed some errors after testing locally.

cowmanjoe committed:
Merged from feature/99 into feature/110.

cowmanjoe committed:
WIP continues.

cowmanjoe committed:
Got typeorm to return the database entities.

cowmanjoe committed:
Ported search with keyword and location filtering to typeorm.

cowmanjoe committed:
Ported search API to typeorm.

cowmanjoe commented:
Also, the database I have with all the sample data can be generated by using the command `psql db_name < internado.out` where db_name is the name of your database on your local instance and internado.out is the path to this [pg_dump file](https://mega.nz/#!09s2BKpY!aZnN6NMnikhmuSq33vTo_nPaLq37t3rJ8pQs03MqoOk).

cowmanjoe committed:
Cleaned up code a bit.

frostyshadows commented on server/models/job.ts:
```diff
@@ -1,12 +1,54 @@
+    public description: string;
+
+    @Column({ nullable: true, length: 256})
+    public city: string;
```
I just added a "state" column to the proposed schema, since a country could have a few cities with the same name. Could you add it as part of this PR? https://github.com/ubclaunchpad/Internado/wiki/Proposed-Database-Schema

frostyshadows commented on server/controllers/searchController.ts:
```diff
@@ -1,38 +1,23 @@
-import { Client } from "pg";
+import {Request, Response} from "express";
+import {Connection, createConnection, getConnection, SelectQueryBuilder} from "typeorm";
```
createConnection isn't actually used, so it doesn't have to be imported

frostyshadows commented on server/controllers/searchController.ts:
```diff
@@ -57,15 +42,21 @@ function getSearchRequest(req: Request, res: Response): SearchRequest {
+        radius = null;
+    } else if ((typeof radius !== "number") || (typeof latitude !== "number") || (typeof longitude !== "number") ||
+        latitude < -90 || latitude > 90 || longitude < -180 || longitude > 180) {
+        res.status(400).send("If radius is defined radius, longitude and latitude must be valid.");
```
I think this should be `If radius is defined radius, longitude and latitude must be valid.`

frostyshadows reviewed (changes requested):
haven't tested it thoroughly yet, but looks good so far! left a few comments

cowmanjoe committed:
Modified schema to match the one on the wiki.

cowmanjoe committed:
Fixed incorrect string and unused import.

cowmanjoe committed:
Made error messages return as JSON.

cowmanjoe commented:
I've updated the schema in my code to match the [proposed schema](https://github.com/ubclaunchpad/Internado/wiki/Proposed-Database-Schema) in the wiki. Also updated the [pg_dump file](https://mega.nz/#!1hlwVaKQ!17pMbyhXCCb3EiojIqhP6UQdT9N_M0kFz-3fDEd7fxU) so it matches that schema. To use this file on your local machine, install postgres and create a database. Once that database is running run `psql {db_name} < internado.sql` in the directory where the internado.sql file I've linked is located.

kimoantiqe commented:
@cowmanjoe Shall I update my PR ( #118) with the above new pg dump ?

cowmanjoe commented:
@kimoantiqe Yup, sounds good.
